server/src/main/java/io/crate/execution/dml/TranslogIndexer.java

```diff
@@ -73,15 +80,34 @@ public TranslogIndexer(DocTableInfo table, Version shardCreatedVersion) {
                 sourceParser.register(DocReferences.toDocLookup(ref).column(), ref.valueType());
             }
         }
-        for (var ref : table.indexColumns()) {
-            for (var source : ref.columns()) {
-                tableIndexSources.computeIfAbsent(source.column().fqn(), _ -> new ArrayList<>()).add(ref.storageIdent());
-            }
-        }
+
+        this.indexColumns = buildIndexColumns(table.indexColumns(), table::getReference, reference -> new TranslogInput(reference.column()));
         ignoreUnknownColumns = table.columnPolicy() != ColumnPolicy.STRICT;
         this.shardCreatedVersion = shardCreatedVersion;
     }
 
+
+    private static class TranslogInput implements Input<Object> {
+
+        private final ColumnIdent ident;
+        private Object value;
+
+        TranslogInput(ColumnIdent ident) {
+            this.ident = ident;
+        }
+
+        @SuppressWarnings("unchecked")
+        public void prepareValue(Object value, DataType<?> dataType) {
+            this.value = dataType.id() != ObjectType.ID ? value
+                : Maps.getByPath((Map<String, Object>) value, ident.path());
+        }
+
```

Comment:
Not sure if we already have some expression in place to get rid of this . I think we don't have an expression that has type dependant `setNextRow`.